File: Day10/Java/UserServiceRefactor.java
// File: UserServiceRefactor.java
// Topic: Structs vs Java Classes — Side-by-side Refactor Example (Java Version)
//
// Scenario: Register a user, prevent duplicates, and send a welcome notification.
//
// Key Java traits:
// - Class bundles state + behavior
// - Explicit encapsulation via private/public
// - Exceptions often used (here we keep checked/unchecked simple)
//
// NOTE: This is a simplified, framework-free example.

import java.util.HashMap;
import java.util.Map;

public class UserServiceRefactor {

    // --- Domain model (class = data + behavior + lifecycle) ---
    static class User {
        private final String name;
        private String email;

        public User(String name, String email) {
            if (name == null || name.trim().isEmpty()) throw new IllegalArgumentException("name empty");
            if (email == null || !email.contains("@")) throw new IllegalArgumentException("email invalid");
            this.name = name.trim();
            this.email = email.trim();
        }

        public String getName() { return name; }
        public String getEmail() { return email; }

        public void setEmail(String email) {
            if (email == null || !email.contains("@")) throw new IllegalArgumentException("email invalid");
            this.email = email.trim();
        }
    }

    // --- Interface + explicit implements ---
    interface Notifier {
        void notify(User to, String message);
    }

    static class EmailNotifier implements Notifier {
        @Override public void notify(User to, String message) {
            System.out.println("Email sent to " + to.getEmail() + ": \"" + message + "\"");
        }
    }

    // --- Repository ---
    interface UserRepo {
        void save(User u);
        User findByEmail(String email);
    }

    static class InMemoryUserRepo implements UserRepo {
        private final Map<String, User> store = new HashMap<>();

        @Override public void save(User u) {
            String key = u.getEmail();
            if (store.containsKey(key)) throw new IllegalStateException("duplicate user: " + key);
            store.put(key, u);
        }

        @Override public User findByEmail(String email) {
            return store.get(email);
        }
    }

    // --- Service class (logic inside class) ---
    static class UserService {
        private final UserRepo repo;
        private final Notifier notifier;

        public UserService(UserRepo repo, Notifier notifier) {
            this.repo = repo;
            this.notifier = notifier;
        }

        public void register(String name, String email) {
            User u = new User(name, email);
            repo.save(u);
            notifier.notify(u, "Welcome!");
        }

        public User getByEmail(String email) {
            User u = repo.findByEmail(email);
            if (u == null) throw new IllegalStateException("user not found: " + email);
            return u;
        }
    }

    // --- Demo ---
    public static void main(String[] args) {
        UserRepo repo = new InMemoryUserRepo();
        Notifier notifier = new EmailNotifier();
        UserService svc = new UserService(repo, notifier);

        svc.register("Deepak", "deepak@example.com");

        try {
            svc.register("Deepak", "deepak@example.com"); // duplicate
        } catch (Exception e) {
            System.out.println("Expected error: " + e.getMessage());
        }

        try {
            svc.getByEmail("missing@example.com");
        } catch (Exception e) {
            System.out.println("Expected error: " + e.getMessage());
        }
    }
}

File: Day10/Go/classvsstruct/user_service_refactor.go
// File: user_service_refactor.go
// Topic: Structs vs Java Classes — Side-by-side Refactor Example (Go Version)
//
// Scenario: Register a user, prevent duplicates, and send a welcome notification.
//
// Key Go traits:
// - Struct = data only
// - Behavior attached via functions/methods (receiver)
// - Interfaces are implicit
// - Errors are returned values (explicit control flow)

package main

import (
	"errors"
	"fmt"
	"strings"
)

type User struct {
	name  string
	email string
}

func NewUser(name, email string) (User, error) {
	name = strings.TrimSpace(name)
	email = strings.TrimSpace(email)

	if name == "" {
		return User{}, errors.New("name empty")
	}
	if !strings.Contains(email, "@") {
		return User{}, errors.New("email invalid")
	}
	return User{name: name, email: email}, nil
}

type Notifier interface {
	Notify(to User, message string) error
}

type EmailNotifier struct{}

func (EmailNotifier) Notify(to User, message string) error {
	fmt.Printf("Email sent to %s: %q\n", to.email, message)
	return nil
}

type UserRepo interface {
	Save(u User) error
	FindByEmail(email string) (User, bool)
}

type InMemoryUserRepo struct {
	store map[string]User
}

func NewInMemoryUserRepo() *InMemoryUserRepo {
	return &InMemoryUserRepo{store: make(map[string]User)}
}

func (r *InMemoryUserRepo) Save(u User) error {
	if _, exists := r.store[u.email]; exists {
		return fmt.Errorf("duplicate user: %s", u.email)
	}
	r.store[u.email] = u
	return nil
}

func (r *InMemoryUserRepo) FindByEmail(email string) (User, bool) {
	u, ok := r.store[strings.TrimSpace(email)]
	return u, ok
}

type UserService struct {
	repo     UserRepo
	notifier Notifier
}

func NewUserService(repo UserRepo, notifier Notifier) *UserService {
	return &UserService{repo: repo, notifier: notifier}
}

func (s *UserService) Register(name, email string) error {
	u, err := NewUser(name, email)
	if err != nil {
		return fmt.Errorf("validation failed: %w", err)
	}
	if err := s.repo.Save(u); err != nil {
		return fmt.Errorf("save failed: %w", err)
	}
	if err := s.notifier.Notify(u, "Welcome!"); err != nil {
		return fmt.Errorf("notify failed: %w", err)
	}
	return nil
}

func (s *UserService) GetByEmail(email string) (User, error) {
	u, ok := s.repo.FindByEmail(email)
	if !ok {
		return User{}, fmt.Errorf("user not found: %s", email)
	}
	return u, nil
}

func main() {
	repo := NewInMemoryUserRepo()
	notifier := EmailNotifier{}
	svc := NewUserService(repo, notifier)

	_ = svc.Register("Deepak", "deepak@example.com")

	if err := svc.Register("Deepak", "deepak@example.com"); err != nil {
		fmt.Println("Expected error:", err)
	}

	if _, err := svc.GetByEmail("missing@example.com"); err != nil {
		fmt.Println("Expected error:", err)
	}
}
